File: problems/greedy_algorithms/valid_starting_city/js/valid_starting_city.js
export const validStartingCity = (distances, fuel, mpg) => {
  const n = distances.length;
  let remainingFuel = 0;
  let startCity = 0;

  for (let i = 0; i < n; i++) {
    remainingFuel += fuel[i] * mpg - distances[i];
    if (remainingFuel < 0) {
      remainingFuel = 0;
      startCity = i + 1;
    }
  }
  return startCity % n;
};

File: problems/greedy_algorithms/valid_starting_city/cpp/valid_starting_city.cpp
#include "valid_starting_city.h"

int validStartingCity(std::vector<int> distances, std::vector<int> fuel,
                      int mpg) {

  int remainingFuel = 0;
  int startCity = 0;

  for (size_t i = 0; i < distances.size(); i++) {
    remainingFuel += fuel[i] * mpg - distances[i];
    if (remainingFuel < 0) {
      startCity = i + 1;
      remainingFuel = 0;
    }
  }
  return startCity % distances.size();
}
